src/test/metrics_api_test.cc

```diff
@@ -236,30 +344,239 @@ TEST_F(MetricsApiTest, TestGaugeEndToEnd)
       TRITONSERVER_GetMetricKind(metric, &kind_tmp), "query metric kind");
   ASSERT_EQ(kind_tmp, kind);
 
-  // Check metrics via C API
-  ASSERT_NE(server_, nullptr);
-  TRITONSERVER_Metrics* metrics = nullptr;
-  FAIL_TEST_IF_ERR(
-      TRITONSERVER_ServerMetrics(server_, &metrics), "fetch metrics");
-  const char* base;
-  size_t byte_size;
-  FAIL_TEST_IF_ERR(
-      TRITONSERVER_MetricsFormatted(
-          metrics, TRITONSERVER_METRIC_PROMETHEUS, &base, &byte_size),
-      "format metrics string");
-  auto metrics_str = std::string(base, byte_size);
-
   // Assert custom metric is reported and found in output
-  auto found = metrics_str.find(std::string("# HELP ") + name);
-  std::cout << metrics_str << std::endl;
-  ASSERT_NE(found, std::string::npos);
-  std::cout << "======" << std::endl;
-  std::cout << metrics_str.substr(found, std::string::npos) << std::endl;
+  ASSERT_EQ(NumMetricMatches(server_, description), 1);
 
   // Cleanup
   FAIL_TEST_IF_ERR(TRITONSERVER_MetricDelete(metric), "delete metric");
   FAIL_TEST_IF_ERR(
       TRITONSERVER_MetricFamilyDelete(family), "delete metric family");
+
+  // Assert custom metric/family is unregistered and no longer in output
+  ASSERT_EQ(NumMetricMatches(server_, description), 0);
+}
+
+// Test that a duplicate metric family can't be added
+// with a conflicting type/kind
+TEST_F(MetricsApiTest, TestDupeMetricFamilyDiffKind)
+{
+  // Create metric family
+  TRITONSERVER_MetricFamily* family1 = nullptr;
+  TRITONSERVER_MetricKind kind1 = TRITONSERVER_METRIC_KIND_COUNTER;
+  const char* name = "diff_kind_test";
+  const char* description = "diff kind description";
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_MetricFamilyNew(&family1, kind1, name, description),
+      "Creating new metric family1");
+
+  // Create duplicate metric family with different kind
+  TRITONSERVER_MetricFamily* family2 = nullptr;
+  TRITONSERVER_MetricKind kind2 = TRITONSERVER_METRIC_KIND_GAUGE;
+  // Expect this to fail, can't have duplicate name of different kind
+  auto err = TRITONSERVER_MetricFamilyNew(&family2, kind2, name, description);
```

Comment:
If the user's Python implementation have two instances of the model and they are calling `TRITONSERVER_MetricFamilyNew` in their `initialize` function, the initialize call for the second model instance will fail since the metric family already exists. Do you think it makes sense to return the existing metric family instead of returning an error?

Reply:
If two pb instances call `MetricFamilyNew(same_name, same_kind, ...)`, it will succeed and they will get the exiting metric family back.

This test is checking if they try to do `family1 = MetricFamilyNew(same_name, kind=COUNTER)` and `family2 = MetricFamilyName(same_name, kind=GAUGE)`

It is not allowed to have duplicate names of different `kind`

Reply:
Got it. Thanks for the clarification. Resolving.